### PR TITLE
feat: handle sdk transaction in runtime

### PR DESCRIPTION
**Pull request by vuvth**
## Proposed Changes

+ support execute asa deploy, app deploy, transfer algo (and more...) in sdk tx.
+ Move convert EncTx to ExecParams from itxn.ts to ixn.ts.
+ Convert `appl` transaction to ExecParams(DeployApp).
+ Fix: Decimals field should default 0 if we undefined it.
+ Fix: metadatahash(ASA Deploy) should convert back to string when we convert EncTx to ExecParams.
+ Test for those cases.

Please review it. Thank you <3

### CODE REVIEW

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 1]
```diff
@@ -3920,8 +3920,25 @@ export class ITxnSubmit extends Op {
       );
     }
 
+    // initial contract account.
+    const appID = this.interpreter.runtime.ctx.tx.apid;
+    const contractAddress = getApplicationAddress(appID as number);
+    const contractAccount = {
+      addr: contractAddress,
+      sk: Buffer.from([])
+    };
```
Comment (ratik21) [inline packages/runtime/src/interpreter/opcode-list.ts:3929]: can't we use `getAccount` here? I am assuming application is already deployed at this point, so the contract account must be present in runtime's context. right?

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 2]
```diff
@@ -3920,8 +3920,25 @@ export class ITxnSubmit extends Op {
...
+
     // get execution txn params (parsed from encoded sdk txn obj)
-    const execParams = parseEncodedTxnToExecParams(this.interpreter.subTxn, this.interpreter, this.line);
+    // singer will be contractAccount
+    const execParams = encTxToExecParams(
+      this.interpreter.subTxn,
+      {
+        sign: types.SignType.SecretKey,
+        fromAccount: contractAccount
+      },
```
Comment (ratik21) [inline packages/runtime/src/interpreter/opcode-list.ts:3938]: Why not set this inside in `encTxToExecParams`?

File: packages/runtime/src/lib/txn.ts [hunk 3]
```diff
@@ -1,11 +1,13 @@
-import { parsing } from "@algo-builder/web";
-import { EncodedAssetParams, EncodedGlobalStateSchema, Transaction } from "algosdk";
+import { parsing, types } from "@algo-builder/web";
+import { AccountAddress } from "@algo-builder/web/build/types";
```
Comment (ratik21) [inline packages/runtime/src/lib/txn.ts:2]: import from build

File: packages/runtime/src/lib/txn.ts [hunk 4]
```diff
@@ -191,3 +193,178 @@ export function isEncTxAssetConfig (txn: EncTx): boolean {
     (txn.caid !== undefined && txn.caid !== 0) && // assetIndex should not be 0
     !isEncTxAssetDeletion(txn); // AND should not be asset deletion
 }
+
+export function isEncTxApplicationCreate (txn: EncTx): boolean {
+  return txn.type === 'appl' && (txn.apan === 0 || txn.apan === undefined);
```
Comment (ratik21) [inline packages/runtime/src/lib/txn.ts:198]: Instead of `'appl'` use an ENUM. It's more intuitive

File: packages/runtime/src/lib/txn.ts [hunk 5]
```diff
@@ -191,3 +193,178 @@ export function isEncTxAssetConfig (txn: EncTx): boolean {
...
+}
+
+export function transactionAndSignToExecParams (
+  txAndSign: types.TransactionAndSign, ctx: Context, line?: number
+): types.ExecParams {
```
Comment (ratik21) [inline packages/runtime/src/lib/txn.ts:203]: As a good practice let's try to add Description in functions, along with the parameters.

File: packages/runtime/src/runtime.ts [hunk 6]
```diff
@@ -802,22 +803,40 @@ export class Runtime {
...
-        case types.TransactionType.DeployASA: {
-          if (txn.asaDef === undefined) txn.asaDef = this.loadedAssetsDefs[txn.asaName];
-          break;
-        }
-        case types.TransactionType.DeployApp: {
-          txn.approvalProg = new Uint8Array(32); // mock approval program
-          txn.clearProg = new Uint8Array(32); // mock clear program
-          break;
+  executeTx (
+    txnParams: types.ExecParams | types.ExecParams[] | types.TransactionAndSign | types.TransactionAndSign[],
```
Comment (ratik21) [inline packages/runtime/src/runtime.ts:807]: we should create a type for this, and use in both `algob.executeTransaction` & `runtime.executeTx`

File: packages/runtime/src/types.ts [hunk 7]
```diff
@@ -23,6 +23,8 @@ export type TEALStack = IStack<bigint | Uint8Array>;
 
 export interface EncTx extends EncodedTransaction {
   txID: string
+  approvalProgram?: string
+  clearProgram?: string
```
Comment (ratik21) [inline packages/runtime/src/types.ts:27]: Please add a note that we expect TEAL code as a full string here (not the compiled version). Which should we taken as input from the user itself

File: packages/runtime/test/integration/SDKTxnRuntime.ts [hunk 8]
```diff
@@ -0,0 +1,148 @@
+import { types } from "@algo-builder/web";
+import { mkTransaction } from "@algo-builder/web/build/lib/txn";
+import { ASADef } from "@algo-builder/web/build/types";
```
Comment (ratik21) [inline packages/runtime/test/integration/SDKTxnRuntime.ts:3]: imports from build

File: packages/runtime/test/integration/SDKTxnRuntime.ts [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+import { getProgram, loadASAFile } from "../../src";
+import { AccountStore } from "../../src/account";
+import { convertToString } from "../../src/lib/parsing";
+import { mockSuggestedParams } from "../../src/mock/tx";
+import { Runtime } from "../../src/runtime";
+import { AccountStoreI, TxReceipt } from "../../src/types";
+import { useFixture } from "../helpers/integration";
+
+const minBalance = BigInt(1e7);
```
Comment (ratik21) [inline packages/runtime/test/integration/SDKTxnRuntime.ts:15]:
```suggestion
const minBalance = BigInt(10 * 1e6);
```

File: packages/runtime/test/integration/SDKTxnRuntime.ts [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+// TODO: add more test for all transaction types.
```
Comment (ratik21) [inline packages/runtime/test/integration/SDKTxnRuntime.ts:17]: Can we create a task and add a link to it here?

File: packages/runtime/test/integration/SDKTxnRuntime.ts [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+describe("Execute SDK transaction", function () {
```
Comment (ratik21) [inline packages/runtime/test/integration/SDKTxnRuntime.ts:18]:
```suggestion
describe("Execute SDK transaction object using runtime", function () {
```

File: packages/runtime/test/integration/SDKTxnRuntime.ts [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+        fromAccount: alice.account,
+        type: types.TransactionType.DeployASA,
+        asaName,
+        asaDef,
+        payFlags: {
+          totalFee: fee
+        }
+      };
+
+      const txAndSign = mkTransactionAndSign(runtime, execParams);
```
Comment (ratik21) [inline packages/runtime/test/integration/SDKTxnRuntime.ts:81]: How about adding a test directly using `SDK` create methods (eg. `mkAssetTransferTransactionFromSuggestedParams`). I think user will directly use that function and call runtime's executeTx, not going through `execParams` --> `mkTransactionAndSign`. 

But if that changes too much of this test, let's leave as it is.

File: packages/runtime/test/integration/SDKTxnRuntime.ts [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+      runtime.executeTx(txAndSign);
+    });
+
+    it("Check application after deploy", () => {
```
Comment (ratik21) [inline packages/runtime/test/integration/SDKTxnRuntime.ts:124]:
```suggestion
    it("Check application exists after deployment", () => {
```

File: packages/runtime/test/src/lib/txn.ts [hunk 14]
```diff
@@ -0,0 +1,184 @@
+import { types } from "@algo-builder/web";
+import { stringToBytes } from "@algo-builder/web/build/lib/parsing";
```
Comment (ratik21) [inline packages/runtime/test/src/lib/txn.ts:2]: import from build

File: packages/runtime/test/src/lib/txn.ts [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+import { encodeBase64 } from "tweetnacl-ts";
+
+import { AccountStore } from "../../../src";
+import { convertToBuffer } from "../../../src/lib/parsing";
+import { encTxToExecParams } from "../../../src/lib/txn";
+import { Runtime } from "../../../src/runtime";
+import { AccountStoreI, EncTx } from "../../../src/types";
+import { useFixture } from "../../helpers/integration";
+
+describe("Convert encode Tx to ExecParams", function () {
```
Comment (ratik21) [inline packages/runtime/test/src/lib/txn.ts:14]:
```suggestion
describe("Convert encoded Txn to ExecParams", function () {
```

File: packages/runtime/test/src/lib/txn.ts [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+    const [encTx] = runtime.createTxnContext(execParams);
+    const sign = {
+      sign: types.SignType.SecretKey,
+      fromAccount: execParams.fromAccount
+    };
+    // add approvalProgram and clearProgram to encTx
+    if (execParams.type === types.TransactionType.DeployApp) {
+      encTx.approvalProgram = execParams.approvalProgram;
+      encTx.clearProgram = execParams.clearProgram;
+    }
```
Comment (ratik21) [inline packages/runtime/test/src/lib/txn.ts:38]: I think i saw this code previously. Can we reuse in a function?

File: packages/runtime/test/src/lib/txn.ts [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+        sign: types.SignType.SecretKey,
+        fromAccount: john.account,
+        type: types.TransactionType.DeployASA,
+        asaName: 'gold',
+        payFlags: { totalFee: 1000 }
+      };
+
+      execParams.asaDef = runtime.loadedAssetsDefs[execParams.asaName];
+
+      assertConvertParams(runtime, execParams);
```
Comment (ratik21) [inline packages/runtime/test/src/lib/txn.ts:76]:
```suggestion
      execParams.asaDef = runtime.loadedAssetsDefs[execParams.asaName];
      assertConvertParams(runtime, execParams);
```

File: packages/runtime/test/src/lib/txn.ts [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+  let execParams: types.ExecParams;
+  this.beforeEach(() => {
+    john = new AccountStore(1e9);
+    smith = new AccountStore(1e9);
+
+    runtime = new Runtime([john, smith]);
+  });
+
+  // helper - help convert and check param from EncTx to ExecParams
+  function assertConvertParams (runtime: Runtime, execParams: types.ExecParams): void {
```
Comment (ratik21) [inline packages/runtime/test/src/lib/txn.ts:28]:
```suggestion
  function assertEncTxConvertedToExecParam (runtime: Runtime, execParams: types.ExecParams): void {
```

File: packages/runtime/test/src/runtime.ts [hunk 19]
```diff
@@ -1,5 +1,5 @@
 import { types } from "@algo-builder/web";
-import { AccountAddress, AlgoTransferParam } from "@algo-builder/web/build/types";
+import { AlgoTransferParam } from "@algo-builder/web/build/types";
```
Comment (ratik21) [inline packages/runtime/test/src/runtime.ts:2]: Import from build.

**Review by ratik21 — COMMENTED** [review]
Thanks for adding comprehensive tests. We also need:
+ changelog update
+ docs update

File: packages/runtime/test/integration/SDKTxnRuntime.ts (hunk 12 above)
Comment (vuvth) [reply packages/runtime/test/integration/SDKTxnRuntime.ts:81]: > How about adding a test directly using `SDK` create methods (eg. `mkAssetTransferTransactionFromSuggestedParams`). I think user will directly use that function and call runtime's executeTx, not going through `execParams` --> `mkTransactionAndSign`.
> 
> But if that changes too much of this test, let's leave as it is.

First time, I also have same approach like you(use mkAssetTransferTransactionFromSuggestedParams, makeApplicationCreateTxn)  However, I see in method [mkTransaction](https://github.com/scale-it/algo-builder/blob/0650e5283ddf92633dab54ba8888034c0bd56cef/packages/web/src/lib/txn.ts#L57) in `algob/web` we ready create SDK transaction. My idea is proof we can run SDK transaction in Runtime and reuse mkTransaction method help me create SDK Txn easier.

File: packages/runtime/src/interpreter/opcode-list.ts (hunk 2 above)
Comment (vuvth) [reply packages/runtime/src/interpreter/opcode-list.ts:3938]: > Why not set this inside in `encTxToExecParams`?

`encTxToExecParams` more general now. We use it to convert `TransactionAndSign` to `ExecParams`, so `sign` can be Lsig. I think keep it in parameters of method is better.

File: packages/runtime/src/interpreter/opcode-list.ts (hunk 1 above)
Comment (vuvth) [reply packages/runtime/src/interpreter/opcode-list.ts:3929]: Yes, I will use runtime.getAccount in this case.

File: packages/runtime/test/src/lib/txn.ts (hunk 16 above)
Comment (vuvth) [reply packages/runtime/test/src/lib/txn.ts:36]: I think those function only use for test. No idea how to reuse it.

File: packages/runtime/src/runtime.ts (hunk 6 above)
Comment (vuvth) [reply packages/runtime/src/runtime.ts:807]: Can we do it on [this task](https://www.pivotaltracker.com/n/projects/2452320/stories/181295625)? I think it more better.

File: packages/runtime/test/integration/SDKTxnRuntime.ts (hunk 12 above)
Comment (ratik21) [reply packages/runtime/test/integration/SDKTxnRuntime.ts:77]: OK

File: packages/runtime/src/runtime.ts (hunk 6 above)
Comment (ratik21) [reply packages/runtime/src/runtime.ts:807]: it's a one liner. but OK, as you prefer

File: packages/runtime/test/src/lib/txn.ts [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+          totalFee: 1000
+        },
+        assetID: 7,
+        freezeTarget: smith.address,
+        freezeState: true
+      };
+      assertEncTxConvertedToExecParam(runtime, execParams);
+    });
+
+    it("Asset Transfer", () => {
```
Comment (sczembor) [inline packages/runtime/test/src/lib/txn.ts:92]:
```suggestion
    it("Should transfer asset", () => {
```

File: packages/runtime/test/integration/SDKTxnRuntime.ts [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+import { mockSuggestedParams } from "../../src/mock/tx";
+import { Runtime } from "../../src/runtime";
+import { AccountStoreI } from "../../src/types";
+import { useFixture } from "../helpers/integration";
+
+const minBalance = BigInt(10 * 1e6);
+
+// TODO: add more test for all transaction types.
+// https://www.pivotaltracker.com/n/projects/2452320/stories/181383052
+describe("Execute SDK transaction object using runtime", function () {
```
Comment (sczembor) [inline packages/runtime/test/integration/SDKTxnRuntime.ts:14]:
```suggestion
describe("Should execute SDK transaction object using runtime", function () {
```

File: packages/runtime/test/integration/SDKTxnRuntime.ts [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+
+  this.beforeEach(() => {
+    alice = new AccountStore(minBalance * 10n);
+    smith = new AccountStore(minBalance * 10n);
+    runtime = new Runtime([alice, smith]);
+  });
+
+  describe("ASA transaction", function () {
+    useFixture('asa-check');
+    it("Deploy ASA transaction", () => {
```
Comment (sczembor) [inline packages/runtime/test/integration/SDKTxnRuntime.ts:62]:
```suggestion
    it("Should deploy ASA transaction", () => {
```

File: packages/runtime/test/integration/SDKTxnRuntime.ts [hunk 23]
```diff
@@ -0,0 +1,144 @@
...
+    });
+
+    it("Check application exists after deployment", () => {
+      const appInfo = runtime.getAppInfoFromName('counter-approval.teal', 'clear.teal');
+      assert.isDefined(appInfo);
+    });
+  });
+
+  describe("Payment transaciton", function () {
+    it("Transfer ALGO Transaction", () => {
```
Comment (sczembor) [inline packages/runtime/test/integration/SDKTxnRuntime.ts:127]:
```suggestion
    it("Should transfer ALGO transaction", () => {
```

File: packages/runtime/src/lib/txn.ts [hunk 24]
```diff
@@ -187,7 +188,192 @@ export function isEncTxAssetDeletion (txn: EncTx): boolean {
...
+  ctx: Context, line?: number): RuntimeAccountI | undefined => {
+  if (publickey === undefined) { return undefined; }
+  const address = encodeAddress(Uint8Array.from(publickey));
+  const runtimeAcc = ctx.getAccount(
+    address
+  );
+  return runtimeAcc.account;
+};
+
+const _getRuntimeAccountAddr = (publickey: Buffer | undefined,
```
Comment (sczembor) [inline packages/runtime/src/lib/txn.ts:371]: same as above (wrong naming convention)

**Review by sczembor — APPROVED** [review]

File: packages/runtime/src/lib/txn.ts [hunk 25]
```diff
@@ -50,7 +51,7 @@ export function parseToStackElem (a: unknown, field: TxField): StackElem {
  * https://github.com/algorand/js-algorand-sdk/blob/e07d99a2b6bd91c4c19704f107cfca398aeb9619/src/transaction.ts#L528
  */
 export function checkIfAssetDeletionTx (txn: Transaction): boolean {
-  return txn.type === 'acfg' && // type should be asset config
+  return ((txn.type as string) === EncTxnType.acfg) && // type should be asset config
```
Comment (robert-zaremba) [inline packages/runtime/src/lib/txn.ts:54]: let's don't use `as`, we already have many problems with it.

**Review by robert-zaremba — APPROVED** [review]

File: CHANGELOG.md [hunk 26]
```diff
@@ -23,6 +23,7 @@ Added:
 Changed:
 - `bond-token-flow` test to also use runtime.defaultAccounts. (see [example](https://github.com/scale-it/algo-builder/blob/develop/examples/bond/test/bond-token-flow.js))
 - The `compile.ts` has been updated and now the tealCode is stored in cache when `scTmplParams` are used to compile TEAL with hardcoded params.
+- Support execution of `transactionAndSign` in Runtime [#601](https://github.com/scale-it/algo-builder/pull/601).
```
Comment (robert-zaremba) [inline CHANGELOG.md:26]:
```suggestion
- Support execution of algo-sdk-js `transactionAndSign` in Runtime [#601](https://github.com/scale-it/algo-builder/pull/601).
```

**Pull request merged**